docs/broker/filtering.md

```diff
@@ -0,0 +1,336 @@
+# Advanced Filtering for Triggers Proposal
+
+## Problem
+
+As an `Event Consumer` I want to filter events in triggers with a variety of
+strategies beyond exact match on predefined fields. Filtering strategies that
+have been requested include:
+
+*   Filtering by source or type prefix or regular expression
+*   Filtering by a custom extension field
+*   Filtering by data fields
+
+The existing filter capability of exact match on source and type is simple to
+understand and use, but doesn't support these more advanced filtering scenarios.
+
+## Objective
+
+Design a flexible filtering mechanism that supports all of the advanced use
+cases identified and can also implement simpler cases like exact match.
+
+The filtering mechanism MUST be safe and efficient. It SHOULD be compatible with
+a future
+[upstream propagation solution](https://github.com/knative/eventing/issues/934)
+for delivery efficiency at scale.
+
+Filters should be easy for the user to specify and simple to interpret at a
+glance.
+
+## Requirements
+
+*   Support all of the advanced filtering use cases identified.
+*   Can implement simpler cases like exact match.
+*   Safe and secure when embedded in multi-tenant processes.
+*   Compatible with upstream filter propagation.
+*   Easily specified by the user.
+*   Simple for users to interpret.
+
+## Proposed Solution
+
+Add a field to the Trigger Filter spec allowing users to specify an expression
+as a string. The expression is evaluated for each event considered by the
+Trigger. If the expression evaluates to true, the event is delivered to the
+Trigger's subscriber.
+
+The evaluation environment of the expression may include:
+
+*   Standard CloudEvent fields like type and source
+*   Custom extensions dynamically parsed from the event
+*   Data fields parsed from the event, if the data content type is known to be
+    parseable
+
+Since parsing dynamic extension and data fields may be costly, the Trigger must
```

Comment:
I'm vary wary of exposing standard and custom attributes in different ways, because 0.2's custom attribute may become 0.3's standard attribute.

If we go down this path, then we need to pick a list of standard attributes, not rely on whatever the current CloudEvents spec says.

---

Another option is to put _all_ attributes into the custom attribute section. If users want to use a standard attribute, then they can get it either from the standard location or the custom location. That way as attributes are made standard, older filters can continue to use them unaltered. 

*Note*: This would not help if a standard attribute is later removed.

Reply:
Currently the variable prefixing is due to a limitation in the CloudEvents SDK. I explain this further at https://github.com/knative/eventing/pull/1047/files#diff-c996c6785529ed44b57aa2d76753453eR219, but that section probably wasn't written when you wrote this comment.

In short, the SDK has no obvious way to combine the official CloudEvents fields into the same map as the extension fields. I hope to discuss this with @n3wscott soon.

Since I expect this to be fixed, I'm happy to change the variable naming scheme to reflect how we hope the SDK will work.

The more I think about it, the more I prefer this scheme:

* `ce`: Official CloudEvents fields defined in the spec, PLUS extensions if they have been parsed. Extensions that conflict with the spec are discarded.
* `ext`: Extensions parsed dynamically from the event, NOT official CloudEvents fields. Extensions that conflict with the spec are retained.
* `data`: Fields parsed dynamically from the event data.

Currently the JSON transport mechanism doesn't allow extensions to conflict with official fields, but it's not clear to me that that property will hold for all transports.

Reply:
I updated the proposal to eliminate `ext`. Extension attributes, when parsed, are now co-mingled with official attributes in `ce`.

> Currently the JSON transport mechanism doesn't allow extensions to conflict with official fields, but it's not clear to me that that property will hold for all transports.

While this is theoretically a concern, it's not something we need to worry about today. If CloudEvents does eventually allow extension attributes with the same name as official attributes, we can add the `ext` variable then.